python/GafferUI/CodeWidget.py

```diff
@@ -0,0 +1,577 @@
+					elif keyword.iskeyword( string ) or string == "self" :
+						highlightType = self.Type.Keyword
+					else :
+						pendingName = self.Highlight( start[1], end[1], self.Type.Call )
+						continue
+				elif tokenType == token.OP :
+					if string in self.__braces :
+						highlightType = self.Type.Braces
+						if string == "(" :
+							if pendingName is not None :
+								result.append( pendingName )
+					elif string in self.__operators :
+						highlightType = self.Type.Operator
+				elif tokenType == token.STRING :
+					highlightType = self.__stringType( string[-1] )
+				elif tokenType == tokenize.COMMENT :
+					highlightType = self.Type.Comment
+				elif tokenType == token.NUMBER :
+					highlightType = self.Type.Number
+
+				if highlightType is not None  :
+					result.append( self.Highlight( start[1], end[1], highlightType ) )
+
+				pendingName = None
+		except tokenize.TokenError as e :
+			if e.args[0] == "EOF in multi-line string" :
+				result.append( self.Highlight( e.args[1][1], None, self.__stringType( line[e.args[1][1]] ) ) )
+
+		return result
+
+	def __stringType( self, quote ) :
+
+		return {
+			"'" : self.Type.SingleQuotedString,
+			'"' : self.Type.DoubleQuotedString
+		}[quote]
+
+CodeWidget.Highlighter = Highlighter
+CodeWidget.PythonHighlighter = PythonHighlighter
+
+# QSyntaxHighlighter used to adapt our Highlighter class
+# for use with a QTextDocument.
+class _QtHighlighter( QtGui.QSyntaxHighlighter ) :
+
+	def __init__( self, document ) :
+
+		QtGui.QSyntaxHighlighter.__init__( self, document )
+
+		self.__highlighter = None
+
+		def format( color ) :
+
+			f = QtGui.QTextCharFormat()
+			f.setForeground( color )
+			return f
+
+		self.__formats = {
+			Highlighter.Type.SingleQuotedString : format( QtGui.QColor( 216, 146, 115 ) ),
+			Highlighter.Type.DoubleQuotedString : format( QtGui.QColor( 216, 146, 115 ) ),
+			Highlighter.Type.Number : format( QtGui.QColor( 174, 208, 164 ) ),
+			Highlighter.Type.Keyword : format( QtGui.QColor( 64, 156, 219 ) ),
+			Highlighter.Type.ControlFlow : format( QtGui.QColor( 207, 128, 195 ) ),
+			Highlighter.Type.Braces : None,
+			Highlighter.Type.Operator : None,
+			Highlighter.Type.Call : format( QtGui.QColor( 220, 221, 189 ) ),
+			Highlighter.Type.Comment : format( QtGui.QColor( 125, 154, 104 ) ),
+			Highlighter.Type.ReservedWord : format( QtGui.QColor( 200, 0, 0 ) ),
+			Highlighter.Type.Preprocessor : format( QtGui.QColor( 207, 128, 195 ) ),
+		}
+
+	# Our methods
+	# ===========
+
+	def setHighlighter( self, highlighter ) :
+
+		self.__highlighter = highlighter
+		self.rehighlight()
+
+	def getHighlighter( self ) :
+
+		return self.__highlighter
+
+	# Qt methods
+	# ==========
+
+	def highlightBlock( self, text ) :
+
+		self.setCurrentBlockState( -1 )
+		if self.__highlighter is None :
+			return
+
+		previousType = None
+		if self.previousBlockState() != -1 :
+			previousType = Highlighter.Type( self.previousBlockState() )
```

Comment:
It feels a bit more complicated to use previousBlockState instead of currentBlockState?  We could just move this above the setCurrentBlockState( -1 ), and use currentBlockState to calculate previousType?

Reply:
I don't think that's equivalent. Each block (line of text) has its own state, and `previousBlockState()` returns the state for the block before the one we're being asked to highlight. `currentBlockState()` is independent of that - it's the state for the current block. In other words, `previousBlockState()` means "previousBlock state" not "previous blockState".